fix termination of `openssl speed` in both fips and non-fips

KazKobara opened the pull request:
apps/speed.c: fix termination of `openssl speed` in both fips and non-fips

Skip unsupported/failure algorithms. 

Fixes: #27679

While the following error still appears even after this pull request is adopted, the error will automatically disappear after #27108 is fixed.

```console
Error while initializing signing data structs for ML-DSA-44.
```

KazKobara committed:
apps/speed.c: refactoring by changing redundunt operations

- EVP_KEM_get0_name(kem) to kem_name
- OSSL_NELEM(rsa_choices) to RSA_NUM

KazKobara committed:
apps/speed.c: fix termination of `openssl speed` in both fips and non-fips

Skip unsupported/failure algorithms.

Fixes: #27679

paulidale commented on apps/speed.c:
```diff
@@ -2256,13 +2265,58 @@ int speed_main(int argc, char **argv)
         const char *sig_name = EVP_SIGNATURE_get0_name(s);
 
+        /* skip signature-algorithms for sigs_doit */
+        if (0 /* 0 is to avoid missing/redundant "||" after list selection */
```
Rather than this ugly, why not start with ECDSA which is unconditional and alphabetically earlier than ED?

KazKobara replied:
Thank you for your review comment.
I have noticed that `./Configure no-ec; make test` fails even before applying this pull request.
So, I am considering adding in speed.c
`#if !definded(OPENSSL_NO_EC) || !definded(OPENSSL_NO_ECDSA)` and
`#if !definded(OPENSSL_NO_EC) || !definded(OPENSSL_NO_ECDH)`
(since `"ec" => [ qw(ec2m ecdsa ecdh sm2 gost ecx tls-deprecated-ec) ]` in Configure).
In this case, what would you think appropriate for the first entry?

paulidale replied:
RSA-SHA

paulidale commented on apps/speed.c:
```diff
@@ -2023,6 +2023,14 @@ int speed_main(int argc, char **argv)
+     */
+    static int fips_module = 0;
+    if (OSSL_PROVIDER_available(NULL, "fips"))
+        fips_module = 1;
```
This might be better done per algorithm.  Just because the FIPS provider is available doesn't mean it's the one that algorithms are fetched from.  FIPS + default is a common combination for users.

More specifically: it would be better to try fetching each algorithm and checking for a failure.

t8m replied:
Yeah, instead of hardcoding fips knowledge, I think it would be much better to just try fetching and if it fails, skip it.

KazKobara replied:
I agree "try fetching" approach is ideal.
(Previous committers to speed.c seem to try to take this approach, but left the message "b/o setup is a pain".)
In this workaround, I am going to leave the approach as TODO comments.

I am considering changing (!fips_module) to (default_module)
to clarify in what reason `openssl speed` is terminated when only the module or the modules are used.

KazKobara replied:
I am going to commit a code for the "try fetching" approach using error codes. 
For the current code using `fips_module`, which would you think is better:
- to remove
- to move to the area that is executed if `SKIP_KNOWN_FAILING_ALGS_IN_ADVANCE` is defined (for debugging or maintenance in case of unintended failures)?

KazKobara committed:
apps/speed.c: add OPENSSL_NO_EC to deal with 'Configure no-ec'

KazKobara committed:
apps/speed.c: add a code to skip unsupported or failed algorithms adaptively

- add a code of "try fetching and if it fails, skip it" approach using error codes.
- move the code for skipping known failing algorithms in the area that is executed
  only if SKIP_KNOWN_FAILING_ALGS_IN_ADVANCE is defined.

Fixes: #27679

KazKobara committed:
apps/speed.c: fix coding style

KazKobara committed:
apps/speed.c: change error codes from binary constants to names

- change the binary constant error codes to the corresponding error names (and move the binary constants to the comment area because error messages include them)
- fix typo

Fixes: #27679

KazKobara commented on apps/speed.c:
```diff
@@ -1984,47 +2040,64 @@ int speed_main(int argc, char **argv)
+    if (OSSL_PROVIDER_available(NULL, "default"))
+        default_module = 1;
+#endif /* SKIP_KNOWN_FAILING_ALGS_IN_ADVANCE */
+
```
The code there is currently for maintenance purposes and is executed only when `SKIP_KNOWN_FAILING_ALGS_IN_ADVANCE` is defined.

The area for `SKIP_KNOWN_FAILING_ALGS_IN_ADVANCE` assumes that only one provider is loaded; currently, only either fips or default is loaded. 

For 3rd-party providers, `speed.c` should be used as is (without defining SKIP_KNOWN_FAILING_ALGS_IN_ADVANCE) or the area should be edited according to the 3rd-party provider's behavior.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 216 days ago

KazKobara commented:
> I believe this issue is already resolved. @KazKobara could you please check?

It appears that this issue still persists in 4.0.0-dev (9c32abf), at least.

t8m commented:
This needs a rebase.

KazKobara commented:
> This needs a rebase.

This (for around OpenSSL 3.5) is becoming too obsolete to rebase.
I will prepare a new pull request for OpenSSL 4.

KazKobara commented:
#31101 is the new pull request for OpenSSL 4.

slontis commented:
Can this PR be closed?

paulidale commented:
I think it can be closed.
